### PR TITLE
Convert java module to new API

### PR DESCRIPTION
**Pull request by tristan957**

### CODE REVIEW

File: mesonbuild/modules/java.py [hunk 1]
```diff
@@ -17,60 +17,61 @@
...
+        super().__init__()
         self.methods.update({
             'generate_native_header': self.generate_native_header,
             'generate_native_headers': self.generate_native_headers,
         })
 
-        if 'java' not in interpreter.environment.coredata.compilers[MachineChoice.BUILD]:
-            detect_compiler_for(interpreter.environment, 'java', MachineChoice.BUILD)
-        self.javac = interpreter.environment.coredata.compilers[MachineChoice.BUILD]['java']
+    def __get_java_compiler(self, state: ModuleState) -> Compiler:
```
Comment (tristan957) [inline mesonbuild/modules/java.py:35]: Is there some helper here that I am no aware of?

File: mesonbuild/modules/java.py [hunk 2]
```diff
@@ -17,60 +17,64 @@
...
-from . import ExtensionModule, ModuleReturnValue, ModuleState
-from ..interpreter import Interpreter
+from . import NewExtensionModule, ModuleReturnValue, ModuleState
 
-class JavaModule(ExtensionModule):
+class JavaModule(NewExtensionModule):
     @FeatureNew('Java Module', '0.60.0')
-    def __init__(self, interpreter: Interpreter):
-        super().__init__(interpreter)
+    def __init__(self, *args: T.Any, **kwargs: T.Any):
```
Comment (dcbaker) [inline mesonbuild/modules/java.py:28]: just `def __init__(self) -> None:` right? (need the None to tell mypy to type check the function

File: mesonbuild/modules/java.py [hunk 3]
```diff
@@ -17,60 +17,64 @@
...
+        if 'java' not in state.environment.coredata.compilers[MachineChoice.BUILD]:
+            detect_compiler_for(state.environment, 'java', MachineChoice.BUILD)
+        return state.environment.coredata.compilers[MachineChoice.BUILD]['java']
 
     @FeatureDeprecated('java.generate_native_header', '0.62.0', 'Use java.generate_native_headers instead')
-    @typed_pos_args('generate_native_header', (str, FileHolder))
+    @typed_pos_args('generate_native_header', (str, FileHolder), min_varargs=1, max_varargs=1)
     @typed_kwargs('java.generate_native_header', KwargInfo('package', str, default=None))
     def generate_native_header(self, state: ModuleState, args: T.Tuple[T.Union[str, FileHolder]],
                                kwargs: T.Dict[str, T.Optional[str]]) -> ModuleReturnValue:
```
Comment (dcbaker) [inline mesonbuild/modules/java.py:44]: There are two things that look funny here, the varargs (I don't see any varargs, just one positional argument. and the `str | FileHolder`. I think that should be `str | File`, the holder should already be removed by that point.

**Review by dcbaker — CHANGES_REQUESTED** [review]

File: mesonbuild/modules/java.py (hunk 3 above)
Comment (tristan957) [reply mesonbuild/modules/java.py:44]: Oh shit.... Thanks for your review.

File: mesonbuild/modules/java.py (hunk 2 above)
Comment (tristan957) [reply mesonbuild/modules/java.py:28]: I believe we still receive the interpreter here. Otherwise Python would error out, right?

Comment (dcbaker) [reply mesonbuild/modules/java.py:27]: We shouldn't be using `T.Any` if we can avoid it, its a step in the wrong direction either way. Either leave it as-is, or remove the interpreter argument if we don't need it.

**Review by dcbaker — APPROVED** [review]

**Pull request merged**